Add more functionality to the “Focus Tests” exercise solution

chasenlehara opened the pull request:
When the form is submitted, the component will now:

- Show error messages for invalid fields
- Give focus to the “success” message if the fields are all valid

I think I will probably break this up into multiple exercises, but for now I want to get just the solution code reviewed.

I also ran Prettier on the code to bring some consistency to it. 😅

Solution: https://codesandbox.io/p/sandbox/github/bitovi/trainings/tree/accessible-error-messages/incorporating-accessibility-into-your-dev-process/focus-tests/solution?file=src/App.tsx

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.test.tsx:
```diff
@@ -5,37 +5,122 @@ Exercise: finish the “gives focus to the first invalid input” test below.
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
+
+    await waitFor(() => {
```
the `findBy` api would simplify this a bit, its basically a wrapper around `waitFor` with a `getBy...`

IIRC it recommended to use it instead of the `waitFor` `getBy...` but i can't find the document that says that

https://testing-library.com/docs/dom-testing-library/api-async/#findby-queries

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.tsx:
```diff
@@ -1,58 +1,117 @@
+import React, { FormEvent, useId, useRef, useState } from "react";
+
+interface FormState {
+  firstName: string | undefined;
```
No sure if this matters but i think its worth at least pointing out, there's a difference between `{firstName: string | undefined}` vs `{firstName?: string}`, `string | undefined` requires the key to be explicitly set to `undefined`. 

```ts
type A = {firstName: string | undefined}
type B = {firstName?: string}

const a: A = {firstName: undefined}
// errors
const aa: A = {}

const b: B = {firstName: undefined}
// doesn't error
const bb: B = {}
```

Maybe you want this, maybe you don't. just thought i'd bring it up incase there's instances below where you didn't want to have to explicitly say `undefined`

chasenlehara replied:
Thanks for calling this out. I think I want what I have because I don’t want parts of the state to be updated—all of the state should be updated at once.

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.tsx:
```diff
@@ -1,58 +1,117 @@
-    </form>
+    <>
+      {submittedFormState?.isValid ? (
+        <p ref={setSuccessRef} tabIndex={0}>
```
Why is this in the `tabIndex` if there's not an interaction? (If this is in the training content, don't worry about a long reply I'll read it)

chasenlehara replied:
It’s training content. 😃

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.tsx:
```diff
@@ -1,58 +1,117 @@
+                    ? !firstNameRef.current?.validity.valid
+                    : undefined
+                }
+                id="firstName"
```
Any reason to not use `useId` for these (first and last name) as well?

chasenlehara replied:
I guess I don’t need them at all, so I’ll just remove them.

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.tsx:
```diff
@@ -1,58 +1,117 @@
+          <p>
+            <label>
+              First name:
+              <input
```
I think these inputs need an `aria-errormessage` on them

`aria-errormessage={ hasError ? firstNameErrorId : undefined}`

(pretend `hasError` exists)

chasenlehara replied:
Maybe I’ll cover this in my training. `aria-errormessage ` [doesn’t have great support](https://a11ysupport.io/tech/aria/aria-errormessage_attribute) and it doesn’t accomplish anything more than `aria-invalid` and `aria-describedby` combined.

DavidNic11 commented on incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/index.tsx:
```diff
@@ -2,11 +2,12 @@ import React from "react";
-    console.error("No main element");
-}
\ No newline at end of file
+  console.error("No main element");
```
this log needed?

chasenlehara replied:
Yes, there should be a log when there’s an error.

chasenlehara committed:
Add more functionality to the “Focus Tests” exercise solution

When the form is submitted, the component will now:

- Show error messages for invalid fields
- Give focus to the “success” message if the fields are all valid

I think I will probably break this up into multiple exercises, but for now I want to get just the solution code reviewed.
